### ISSUE TITLE
NLL: unsound verification of higher ranked outlives bounds

### ISSUE THREAD

**Issue by aliemjay** [rust-lang/rust#98095]
This code now passes under under NLL after #95565:  ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e4d68f7aec3814820e945ad4d0253a7c))
```rust
fn assert_static<T>() where for<'a> T: 'a, {}

// `test` passes without requiring `T: 'static`
fn test<T>() { assert_static::<T>(); }
```

`assert_static` should be able to infer that `T: 'static` because, in the language of Chalk, I believe the bound `for<'a> T: 'a` should be lowered to `forall<'a> { Outlives(T: 'a) :- WellFormed(T). }` and because the well-formedness of T is assumed to be true, we can instantiate `'a` to any lifetime including `'static`.

@rustbot label T-compiler T-types regression-from-stable-to-nightly C-bug

**Comment by steffahn** [comment]
```rs
use std::any::Any;

trait Lt<'a>: 'a {}
impl<'a, T: ?Sized + 'a> Lt<'a> for T {}

trait Tr {}
impl<T: ?Sized> Tr for T {}

fn test<T, R: Any>(x: T) -> R {
    // works despite non-'static
    step1::<T, R>(x)
}

// remark: starting in `test` with `step2` directly works, too.
fn step1<T, R: Any>(x: T) -> R
where
    for<'a> T: 'a,
{
    step2::<T, R>(x)
}

fn step2<T: for<'a> Lt<'a>, R: Any>(x: T) -> R {
    step3::<T, R>(x)
}

fn step3<T: Lt<'static>, R: Any>(x: T) -> R {
    step4::<T, R>(x)
}

fn step4<T: 'static, R: Any>(x: T) -> R {
    *(Box::new(x) as Box<dyn Any>).downcast::<R>().unwrap()
}

fn main() {
    let x = String::from("Hello World");
    let r: &'static str = test(&x[..]);
    drop(x);
    println!("{r}");
}
```
```
����0V��
```
([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=3cfbde59f0355491c7ec7df8d2e4d35e))

@rustbot label unsound

**Comment by oli-obk** [comment]
> but calling `step2` from `step1` doesn’t work on stable either, so maybe a `for<'a> T: 'a` constraint on its own is actually kinda useless at the moment?

Yea `fn foo<T>() where for<'a> T: 'a { foo::<T>() }` fails to compile on stable right now, because higher kinded outlives bounds don't satisfy themselves

**Comment by lqd** [comment]
This is used to be rejected, and started being accepted in `nightly-2020-07-28` ([commit range](https://github.com/rust-lang/rust/compare/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f...76e83339bb619aba206e5590b1e4b813a154b199)).

<details>
<summary>List of merged PRs for that nightly</summary>

- #74792 - flip1995:clippyup, r=Manishearth
- #74708 - kanru:issue-74564, r=davidtwco
- #72121 - Aaron1011:final-hygiene-rebase, r=petrochenkov
- #74784 - anp:track-vtables, r=eddyb
- #74653 - petrochenkov:pmenv, r=dtolnay
- #74737 - smmalis37:astconv-factor, r=davidtwco
- #74817 - JohnTitor:rollup-0fchdye, r=JohnTitor, rolling up:
  - #74088 (Avoid writes without any data in `Write::write_all_vectored`)
  - #74598 (Fix sync_once_cell_does_not_leak_partially_constructed_boxes)
  - #74750 (Clean up some uses of logging in ui tests)
  - #74783 (python codes cleanup)
  - #74790 (Don't italicize comments in ayu theme)
  - #74799 (Fixed typo in `closure`)
- #74775 - RalfJung:miri-alloc-ids, r=oli-obk
- #74831 - Manishearth:rollup-ugw4pt4, r=Manishearth, rolling up:
  - #73858 (Make more primitive integer methods const)
  - #74487 (Forbid generic parameters in anon consts inside of type defaults)
  - #74803 (rustbuild: fix bad usage of UNIX exec() in rustc wrapper)
  - #74822 (More ensure stack to avoid segfault with increased `recursion_limit`)
- #73503 - lcnr:forall-predicate-what-and-why-2, r=nikomatsakis

</details>

Out of those, #73503 looks like the most likely candidate, cc PR author @lcnr and reviewer @nikomatsakis.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2398095.20NLL.3A.20unsound.20verification.20of.20higher.20ranked.20outlives.20.E2.80.A6)).

@rustbot label -I-prioritize +P-critical

**Comment by nikomatsakis** [comment]
Discussed with @oli-obk and @lcnr -- we believe the problem is that the outlives code isn't handling placeholder regions correctly... they are in this fallthrough

https://github.com/rust-lang/rust/blob/edab34ab2abbafc16a78daedf71dbacd2eb0b7bf/compiler/rustc_infer/src/infer/outlives/components.rs#L180

but they should be handled more like [parameters](https://github.com/rust-lang/rust/blob/edab34ab2abbafc16a78daedf71dbacd2eb0b7bf/compiler/rustc_infer/src/infer/outlives/components.rs#L117-L121), except that (at least at present) we can only bound them by `'static`. 

I might be able to make this fix today else could mentor someone potentially (ping me on Zulip).

**Comment by nikomatsakis** [comment]
So @compiler-errors and I did a deep dive here. What we found is that:

* There is a bug in the NLL type check (which @compiler-errors described above) that is causing it to be unsound.
* However, fixing that bug exposes a (pre-existing) bug around how we handle `for<'a> T: 'a` bounds, particularly with projections, causing regressions. 
* 
* The good news is that I think fixing this bug will benefit @oli-obk and others.  (I'll describe it in detail later.) The **bad news** is that a proper fix is non-trivial, so for the short term it may be better to revert NLL stabilization. I'll have to check with @jackh726 about how difficult that will be.
